internal/controller/pkg/resolver/reconciler.go

```diff
@@ -375,3 +401,95 @@ func (r *Reconciler) findDependencyVersion(ctx context.Context, dep *v1beta1.Dep
 
 	return addVer, nil
 }
+
+// UpdatableVersionFinder is an implementation of versionFinder that allows for version upgrade/downgrade capability considering parent constraints.
+type UpdatableVersionFinder struct {
+	fetcher xpkg.Fetcher
+	config  xpkg.ConfigStore
+}
+
+// FindValidDependencyVersion finds a valid version with version upgrade/downgrade capability considering parent constraints.
+func (u *UpdatableVersionFinder) FindValidDependencyVersion(ctx context.Context, dep *v1beta1.Dependency, ref name.Reference, n dag.Node, log logging.Logger) (string, error) {
+	if digest, err := conregv1.NewHash(dep.Constraints); err == nil {
+		log.Debug("package is pinned to a specific digest, skipping resolution")
+		return digest.String(), nil
+	}
+
+	ic, ps, err := u.config.PullSecretFor(ctx, ref.String())
+	if err != nil {
+		log.Info("cannot get pull secret from image config store", "error", err)
+		return "", errors.Wrap(err, errGetPullConfig)
+	}
+
+	var s []string
+	if ps != "" {
+		log.Debug("Selected pull secret from image config store", "image", ref.String(), "imageConfig", ic, "pullSecret", ps)
+		s = append(s, ps)
+	}
+	// NOTE(hasheddan): we will be unable to fetch tags for private
+	// dependencies because we do not attach any secrets. Consider copying
+	// secrets from parent dependencies.
+	tags, err := u.fetcher.Tags(ctx, ref, s...)
+	if err != nil {
+		log.Debug(errFetchTags, "error", err)
+		return "", errors.New(errFetchTags)
+	}
+
+	vs := []*semver.Version{}
+	for _, r := range tags {
+		v, err := semver.NewVersion(r)
+		if err != nil {
+			// We skip any tags that are not valid semantic versions.
+			continue
+		}
+		vs = append(vs, v)
+	}
+
+	var cv *semver.Version
+
+	// If the node is a LockPackage, it means it's already installed.
+	if lp, ok := n.(*v1beta1.LockPackage); ok {
+		log.Debug("dependency found in lock, trying to find the minimum valid version which is greater than the current version")
+		sort.Sort(semver.Collection(vs))
+		cv = semver.MustParse(lp.Version)
+	}
+
+	// If the node is not a LockPackage, it means it's a new dependency that we need to install with the latest version.
+	if cv == nil {
+		log.Debug("dependency not found in lock, trying to find the maximum valid version")
+		sort.Sort(sort.Reverse(semver.Collection(vs)))
+	}
+
+	var addVer string
+	for _, v := range vs {
+		valid := true
+		for _, constraint := range dep.GetParentConstraints() {
+			c, err := semver.NewConstraint(constraint)
+			if err != nil {
+				log.Debug(errInvalidConstraint, "error", err)
+				return "", errors.New(errInvalidConstraint)
+			}
+
+			if !c.Check(v) {
+				valid = false
+				break
+			}
+
+			// NOTE(ezgidemirel): If there is no valid version greater than the current version,
+			// we should not allow downgrade.
+			// Compare compares this version to another one. It returns -1, 0, or 1 if
+			// the version smaller, equal, or larger than the other version.
+			if cv != nil && v.Compare(cv) == -1 {
```

Comment:
```suggestion
			if cv != nil && v.LessThan(cv) {
```

Reply:
Wouldn't this cause us to fail incorrectly in the following scenario? Unless I am missing something :)

I have provider-foo `v0.2.0` installed and available tags for this provider as `v0.1.0`, `v0.2.0`, `v0.3.0`, 

I have two packages installed declaring a dependency to provider-foo as `>=v0.1.0` and `>=v0.3.0`. Ideally, it should upgrade it to `v0.3.0`. But when this line is executed for `v`=`v0.1.0` and `c`=`>=v0.1.0`, it will fail since `v0.1.0` is less than the current version.

I believe this check should be somewhere outside of the loop.